mixer/pkg/runtime2/handler/env.go

```diff
@@ -94,3 +97,40 @@ func (e env) ScheduleDaemon(fn adapter.DaemonFunc) {
 		reachedEnd = true
 	}()
 }
+
+func (e env) ensureWorkerClosed() error {
+	if e.counters.daemons != nil {
+
+		m := new(dto.Metric)
+
+		var c prometheus.Metric = e.counters.daemons
+		if err := c.Write(m); err != nil {
+			return e.Logger().Errorf("Failed to fetch adapter's scheduled daemon counter: %v", err)
+
+		} else if *m.GetGauge().Value > 0 {
+			_ = e.Logger().Errorf("Adapter did not close all the scheduled daemons")
+			// TODO: ideally we should return error here so that we can increment the counter that keep track of
+			// error on close that a higher level. However, currently we cannot guarantee that SchedularXXXX gauge
+			// counter will give consistent value because of timing issue in the ScheduleWorker and ScheduleDaemon.
+			// Basically, even if the adapter would have closed everything before returning from Close function, our
+			// counter might get delayed decremented, causing this false positive error.
+			// Therefore, we need a new retry kind logic on handler Close to give time for counters to get updated
+			// before making this as a red flag error. runtime2 work has plans to implement this stuff, we can revisit
+			// this to-do then. Same for the code below related to workers.
+			return nil
+		}
+	}
+
+	if e.counters.workers != nil {
```

Comment:
Can you make this method on envCounters struct?
Also, it looks like m is getting allocated on heap. Any chance we can allocate this on stack?